skimage/future/detect_obj/cascade.pyx

```diff
@@ -0,0 +1,878 @@
+            Py_ssize_t current_width
+            Py_ssize_t current_row
+            Py_ssize_t current_col
+            Py_ssize_t current_step
+            Py_ssize_t number_of_scales
+            Py_ssize_t img_height
+            Py_ssize_t img_width
+            Py_ssize_t scale_number
+            Py_ssize_t window_height = self.window_height
+            Py_ssize_t window_width = self.window_width
+            int result
+            float[::1] scale_factors
+            float[:, ::1] int_img
+            float current_scale_factor
+            vector[Detection] output
+            Detection new_detection
+
+        int_img = self._get_contiguous_integral_image(img)
+        img_height = int_img.shape[0]
+        img_width = int_img.shape[1]
+
+        scale_factors = self._get_valid_scale_factors(min_size, max_size, scale_factor)
+        number_of_scales = scale_factors.shape[0]
+
+        # Initialize lock to enable thread-safe writes to the array
+        # in concurrent loop.
+        cdef openmp.omp_lock_t mylock
+        openmp.omp_init_lock(&mylock)
+
+
+        # As the amount of work between the threads is not equal we use `dynamic`
+        # schedule which enables them to use computing power on demand.
+        for scale_number in prange(0, number_of_scales, schedule='dynamic', nogil=True):
+
+            current_scale_factor = scale_factors[scale_number]
+            current_step = <Py_ssize_t>round(current_scale_factor * step_ratio)
+            current_height = <Py_ssize_t>(window_height * current_scale_factor)
+            current_width = <Py_ssize_t>(window_width * current_scale_factor)
+            max_row = img_height - current_height
+            max_col = img_width - current_width
+
+            # Check if scaled detection window fits in image.
+            if (max_row < 0) or (max_col < 0):
+                continue
+
+            current_row = 0
+            current_col = 0
+
+            while current_row < max_row:
+                while current_col < max_col:
+
+                    result = self.classify(int_img, current_row, current_col, scale_factors[scale_number])
+
+                    if result:
+
+                        new_detection = Detection()
+                        new_detection.r = current_row
+                        new_detection.c = current_col
+                        new_detection.width = current_width
+                        new_detection.height = current_height
+                        openmp.omp_set_lock(&mylock)
+                        output.push_back(new_detection)
+                        openmp.omp_unset_lock(&mylock)
+
+                    current_col = current_col + current_step
+
+                current_row = current_row + current_step
+                current_col = 0
+
+        return list(_group_detections(output, intersection_score_threshold,
+                                      min_neighbour_number))
+
+    def _load_xml(self, xml_file, eps=1e-5):
```

Comment:
@warmspringwinds can you refactor this to use more functions? Every "paragraph" that you start with a comment explaining its purpose is a candidate for refactoring.

NASA [enforces](http://pixelscommander.com/wp-content/uploads/2014/12/P10.pdf) that no function should be longer than 60 lines, including variable declarations (one per line). I think we should stick to that — it allows code review having the full function in sight, rather than having to remember what came before.

This comment also applies to `detect_multi_scale`, above.